### ISSUE TITLE
How to require and work with extensions from the asset portal?

### ISSUE THREAD

**Issue by rama0x9** [hxdefold/hxdefold#27]
hi, it's not mentioned anywhere and me as a total haxe noob have no idea how to do it. 
maybe @:luaRequire('...') or something? 

thanks

**Comment by haath** [comment]
Hey there.

Yes you need to create externs. And for most cases where the extension code is in a lua module `@:luaRequire()` should work.

Here is an example of the extern for [defold-orthographic](https://github.com/britzl/defold-orthographic) from a personal project: https://gitlab.com/haath/dex/-/blob/master/src/dex/lib/orthographic/Camera.hx

I have a few more [here](https://gitlab.com/haath/dex/-/tree/master/src/dex/lib) but I haven't properly tested some of them.

---

If you're having trouble with a specific extension let me know and I could get you started :)

**Comment by rama0x9** [comment]
These are some helpful examples, thanks, 
But unfortunatley not so easy as i thought it would be :) How about native extensions? e.x [gpgs](https://github.com/defold/extension-gpgs). Are they even harder to implement or easier?

**Comment by haath** [comment]
Same thing generally, because even those extensions have a lua API in the end.

From my examples again you can take a look at Astar which is a native extension: https://gitlab.com/haath/dex/-/blob/master/src/dex/lib/astar/Astar.hx

The only difference is I used `@:native()` instead of `@:luaRequire()`, since it's not necessary to require native extensions if I recall correctly.

---

You just need to have a look at the API and replicate just the methods you need in an extern: https://defold.com/extension-gpgs/

For example:

```haxe
@:native("gpgs")
extern class Gpgs
{
    static function login(): Void;

    static function snapshot_set_data(data: String): SnapshotSetDataReturn;
}

@:multiReturn extern class SnapshotSetDataReturn
{
    var success: Bool;
    var error_message: String;
}
```

The only parts that need special attention really are when a lua method returns multiple values, in which case you need to define a `@:multiReturn` class like above.

Then using it in Haxe is straight-forward:

```haxe
Gpgs.login();

var snap: SnapshotSetDataReturn = Gpgs.snapshot_set_data("my data");
```

**Comment by haath** [comment]
And for a little more context since you are new to Haxe, externs are basically just a way to lie to the Haxe compiler and say "a class is going to exist there with those methods, just trust me".

```haxe
@:native("foo")
extern class Foo
{
    static function bar(x: Int): Bool;
}
```

This way, even though the class is not implemented in Haxe, you can write the following without compilation errors:

```haxe
var b: Bool = Foo.bar(5);
```

Which Haxe will then generate as:

```lua
b = foo.bar(5)
```

In the end this is equivalent to just injecting raw lua code like this:

```haxe
untyped __lua__("b = foo.bar(5)");
```

But in a type-safe and cross-platform manner!

**Comment by rama0x9** [comment]
You are awesome man! Thanks for taking your time, more than i asked :)

"**replicate just the methods you need**" this right here is what i wanted to know, kinda like nim.

**Comment by rama0x9** [comment]
More of a general problem with vscode. The code compiles and runs without a flaw, but vscode always complains about it, do you know why? Without the package there is no error.
```haxe
package game;
import defold.support.Script;  <-- Type not found

**Comment by haath** [comment]
Could you please share your `hxml` file, as well as a description of your folder structure?

**Comment by rama0x9** [comment]
build.hxml
```haxe
-cp .
-lua main.lua
-lib hxdefold
-dce full
-D analyzer-optimize
--macro include("", true, null, ["."])
```
structure
```
[project]
| [.git]
| [.vscode]
| [build]
| [content]
| | [game]
| | | | game.collection
| | | | Game.hx
| | [screens]
| | | [map]
| | | | [assets]
| | | | [gui]
| | | | [scenes]
| | | | | map.collection
| | | | | Map.hx
| | [scripts]
| | | [game]
| | | | [scenes]
| | | | | Game.script
| | | [screens]
| | | | [map]
| | | | | [scenes]
| | | | | | Map.script
| | build.hxml
| | game.project
| | main.lua
| [res]
| prj.code-workspace
```

ofc only a fracture of my folders, but that basically shows how i structure my projects. I want to replace former scripts under scenes folder with hx ones.

build.hxml is inside content folder
Game.hx inside content/game

**Comment by haath** [comment]
So, Haxe actually has a strict package-directory relation much like Java.
In your case, your classpath (cp) is `content`, therefore the correct package name for `Game.hx` would be `package game.scenes;`.

And make sure that all of your classes have the correct package names as well.

Also has the vscode Haxe extension correctly identified your `hxml` file? It should be shown on the bottom left like this:

![image](https://user-images.githubusercontent.com/29335643/86814769-23600c00-c082-11ea-8903-4cd72c4fb3ad.png)

At least in my experience, I recall needing to have the `hxml` files in the root of the project for vscode to detect them.

**Comment by rama0x9** [comment]
hxml needs to be in project root, you are right, that solved it :) my project structure i posted was not correct, i changed it, Game.hx is under content/game. But nvm, its working.

k, now i need to get monarch working, wish me luck :)

**Comment by haath** [comment]
Have fun!

**Comment by rama0x9** [comment]
unfortunately i need some more guidance
Defold error: 
```lua
game.lua: module 'monarch' not found
```

I copied both monarch files from your dex repo to my content/lib/ext folder and changed both to `package lib.ext;`

In Monarch.hx `@:luaRequire("monarch")`

In Game.hx `import lib.ext.Monarch;`

I also tried `@:luaRequire("monarch.monarch")` and `@:luaRequire("monarch/monarch")`

What im missing?

Btw, In lua i would do `require "monarch.monarch"` but in your dex example you have `@:luaRequire("monarch")`.
Your dirtylarry example has `@:luaRequire("dirtylarry/dirtylarry")`.

Even if not working, i would say `@:luaRequire("dirtylarry.dirtylarry")`, correct? :)

**Comment by haath** [comment]
Yeah as I said above, some of those externs are untested. I just wrote them while being bored at work one day. 😄 

Looking at Monarch's examples it should indeed be `@:luaRequire("monarch.monarch")`.

What error are you getting with that? 

Also try to find how exactly is it generated inside your `main.lua`. You should be able to spot what is wrong by looking at how what you get in your `main.lua` differs from [britzl's examples](https://github.com/britzl/monarch/tree/master/example).

**Comment by rama0x9** [comment]
regardless of what i put in require, its always type not found in game.lua.

i looked through game.lua code and found nothing suspicious, the line which gives the error is 
```lua
__lib_ext_Monarch = _G.require("monarch.monarch")
```

looking good to me

**Comment by haath** [comment]
And for sure you have the monarch extension added to the project, right?

And you also did `Project > Fetch Libraries` on the Defold editor?

**Comment by rama0x9** [comment]
Sure i did.
Would it be to much to ask if you could take the hxdefold warbattles example and make it work with monarch? Maybe some little popup on startup or so. 
I tried and failed, same error as with my project. Im clearly doing something wrong, probably something really stupid wrong.

I dont know if i should just stick with lua as my project is already quite big. Who knows what other problems i will get into in the long run when trying to port. This surely will take time. 

But Haxe looks really great, damn :)

**Comment by haath** [comment]
So apparently without the `_G` it works...

```lua
__warbattles_monarch_Monarch = require("monarch.monarch")
```

I don't generally write lua, so I'm not sure what this means exactly.

It's strange because the orthographic camera library has the same exact structure as monarch:

```lua
local M = {}

...

return M
```

But for that library this works just fine on my projects:

```lua
__dex_lib_orthographic_Camera = _G.require("orthographic.camera")
```

---

I will try to get some support on this on the Defold forum, meanwhile here is a workaround that should work for you:

Firstly avoid the lua import being generated.

```haxe
@:native("monarch")
extern class Monarch
```

Then patch the file in your build process to prepend the require line manually.

```sh
# Prepend a line to main.lua
sed -i '1s;^;monarch = require("monarch.monarch")\n;' main.lua
```

**Comment by haath** [comment]
fyi https://forum.defold.com/t/difference-between-require-and-g-require-for-asset-libraries/65741/2?u=haath

So from the Haxe side everything is correct, Defold just doesn't recognize `_G.require()` and so the library is excluded from the build.

**Comment by rama0x9** [comment]
Hey, its me again :)  
I need some help with another extern, this time it is [Druid](https://github.com/Insality/druidl).  
```lua
local druid = require("druid.druid")

local function button_callback(self)
	print("Button was clicked!")
end

function init(self)
	self.druid = druid.new(self)
	self.druid:new_button("button_node_name", button_callback)
end

function final(self)
	self.druid:final()
end

function on_input(self, action_id, action)
	return self.druid:on_input(action_id, action)
end
```
How to tame this monster lib? How to "Haxe" this? 
This one is somewhat different then the other "simpler" ones. Can you tell if it's really more difficult to write an extern for it?
E.g. i dont know what the return type is for the `druid.new(self)` function. When i look into druid.lua it returns an instance (it uses middleclass). I think i need to write many more externs for this lib, not only for druid.lua and e.g. button.lua.

**Comment by haath** [comment]
So basically you just need to **not** use `static`. Externs can be normal instance classes as well.

With Lua the difference is something like this:

```haxe
static function foo(): Void;    ->    druid.foo()
```

```haxe
function bar(): Void;           ->    druid:bar()
```

There also exists the metadata `@:luaDotMethod` which can force calls to a method to be generated with `.` instead of `:`, although I don't think you'll need it.

Then `new()` should just be your constructor!

```haxe
@:luaRequire("druid.druid")
extern class Druid
{
    function new();

    // No 'static' keywords means it's an instance method
    function final(): Void;
}
```

Now I'm not sure what exactly that `self` is that is being passed to the constructor, because I haven't used Druid yet myself.
My best guess, by looking at the rest of the Defold API would be that it is the script's `self`, in which case in Haxe that would be your properties type! So with that being true I believe your extern then needs to be generic.


```haxe
@:luaRequire("druid.druid")
extern class Druid<T: {}>
{
    function new(self: T);


    function new_button(button_name: String, callback: T -> Void): Void;
}
```

And then in your script...

```haxe
typedef MyScriptProperties =
{
    @property() var myEditorProperty1: Vector3;
    var myProperty2: Int;
}


class MyScript extends Script<MyScriptProperties>
{
    override function init(self: MyScriptProperties)
    {
        var druid: Druid<MyScriptProperties> = new Druid(self);

        druid.new_button("foo", buttonCallback);
    }

    // Note how we can automatically use MyScriptProperties as the argument here.
    // This is done this way so you can still have a reference to which specific script instance registered the callback above!
    function buttonCallback(self: MyScriptProperties)
    {
 
    }
}
```

Super untested, just writing this on the browser while on vacation, hope it puts you on the right track 😄

**Comment by haath** [comment]
Also regarding the previous `_G.require()` bug, I think there is a *cleaner* workaround than patching your lua file, at least until the Defold team fixes the issue.

Manually create an empty `.script` file and add it to some object in your scene.

```lua
-- dummy.script
-- Add dummy imports for all asset libraries you need to use.
local dummy1 = require("monarch.monarch")
local dummy2 = require("druid.druid")
```

If I understood britzl corectly, this should force the editor to recognize that the libraries are being used and include them in the build.

Then in theory your `@:luaRequire("...")` should work as intended, and when the Defold people fix the issue all your need to do is just remove the dummy script.

**Comment by rama0x9** [comment]
Thank you big times! Hope im not bothering you too much for asking such questions, somewhat embarrassing to be a totally noob :)
Oh btw., as lua modules are non existent in haxe i need to ask if im doing it right. If i want to port a lua module i just use a regular class?  

data.lua
```lua
local share = {foo = {man = 1, choo = 2}, bar = 1}

local function has_mustache()
  return true
end

return {share = share, has_mustache = has_mustache}
```
Data.hx
```haxe
class Data {
  public static var share = ...;
  public static function has_mustache() {
    return true;
  }
}
```
Is this the way to go?

**Comment by rama0x9** [comment]
There are really not enough Haxe<>Lua resources that i could learn from, thats making things a little bit difficult to get going. As you can see in my snippet above, im not even sure what the best equivalent for lua tables in haxe is :D. There must be some "better" ways than json? something more haxish?

**Comment by rama0x9** [comment]
> ```hx
>function final(): Void;
> ```
final is a keyword in haxe, so the compiler complains. (unexpected final)
Not easy to search for a solution when i don't how what to search for, i really tried! :)
I would use something like final_() but how to tell the compiler it should output it as final()

edit:
Ok, i think i need to use metadata. Need to dig in deeper, may i find a solution before your answer  ;) 
edit2:
```hx
@:native("final")
function final_():Void;
```
works

**Comment by haath** [comment]
Ye that's correct. A similar thing was done for the `final()` callback in in the `Script` class.

**Comment by rama0x9** [comment]
In your example above you made druid a local variable, but i need it also outside of scope.
So i did this:
```hx
typedef Self = {
  var druid:Druid<Dynamic>;
  //var button_play: ?Dynamic;
}

class Start extends GuiScript<Self> {
...
  function setup_nodes(self:Self) {
    self.druid = new Druid(self);
    // self.button_play = self.druid.new_button("play", on_button_play_touched);
}
}
```
Now even it it works with Dynamic Type, im not happy with it. What are the proper Types for both names? 
button_play does not work with Dynamic.

**Comment by haath** [comment]
Why do you need Dynamic? Does this not work?

```haxe
typedef Self = {
  var druid:Druid<Self>;
}
```

Also what is `new_button()` supposed to return? You could create a class to match that.

For example if to go according to [here](https://github.com/Insality/druid/blob/master/docs_md/01-components.md#button):

```haxe
extern class Button
{
    function set_click_zone(): Void;
    function set_key_trigger(): Void;
    // etc
}
```

Then `Druid.new_button()` should have a return type `Button`.

**Comment by rama0x9** [comment]
I used Dynamic as a placeholder because i didn't know i can use the structure name itself as a type.

I think im getting the hang of it.  

Thank you for helping me out on these topics, i know you have better things to do :)

**Comment by AsakuraMizu** [comment]
https://github.com/AsakuraMizu/hxdruid

have a try 😄 

(some apis are quite different to make it more "haxeized")

**Comment by Lerg** [comment]
Is it possible to provide haxe mapping for my defold native extension inside the extension for haxe to pick it up? So that users don't need to do anything extra.

**Comment by haath** [comment]
> Is it possible to provide haxe mapping for my defold native extension inside the extension for haxe to pick it up? So that users don't need to do anything extra.

Without doing anything extra, no. Because from what I see Defold keeps downloaded libraries/extensions inside zip files in the `.internal` folder, so it's not possible to reference the contents.

The *fastest* way I assume, would be to instruct your users to extract the zip files and add a directory inside to the classpath with `-cp`. The build/pipeline script can also do this, so it's not a bad idea overall.

Another alternative would be to turn your repo into a haxelib, then your users would add it to haxe as a library. It would only need a `haxelib.json` to point to the classpath of Haxe files.

```
haxelib git https://github.com/Lerg/extension-admob
```

Hope it helps, let me know if you'd like more detailed instructions!

**Comment by adamgoose** [comment]
I was also able to work around the `@:luaRequire` issue by replacing it with `@:native("require('dirtylarry/dirtylarry')")`.

**Comment by haath** [comment]
Release notes for version [1.2.180](https://forum.defold.com/t/defold-1-2-180-has-been-released/67743) mention:

> Issue-5054 3 - Fixed: Improve detection of require() calls in Lua modules and scripts

It sounds like it could fix the problem we've been discussing here but I don't have an ongoing project to test it with at the moment. If some does and could check if it's fixed, please let us know!

**Comment by DavidLyhedDanielsson** [comment]
I'm having some issues too since the major changes to hxdefold. I have a very simple Druid wrapper that just looks like this for now (don't mind the `neww`s and such ;) ):
```
package druid;

import defold.types.*;
import defold.support.*;

extern class DruidInstance {
	@:native("final") function finall():Void;
	function update(dt:Float):Void;
	function on_message<TMessage>(messageId:Message<TMessage>, message:TMessage, sender:Url):Void;
	function on_input(actionId:Hash, action:ScriptOnInputAction):Bool;

	function new_button(node:String, callback:() -> Void):Button;
}

class Button {}

@:luaRequire("druid.druid")
extern class Druid {
	@:native("new") static function neww(context:GuiScript):DruidInstance;
}
```

And I got this to finally work by using `untyped __lua__` which feels wrong:
```
package gui;

import defold.support.GuiScript;
import druid.Druid;
import defold.types.*;
import defold.support.*;

class Main extends GuiScript {
	var druid:DruidInstance;
	var button:Button;

	function onButtonPressed() {
		trace("Button pressed!");
	}

	override function init() {
		super.init();

		this.druid = Druid.neww(**untyped __lua__('_G._hxdefold_self_')**);
		this.button = this.druid.new_button("button_start", onButtonPressed);
	}

	...
}
```

If I just pass `this` to `Druid.neww`, I don't get the correct type. I expect `this` to pass the GuiScript, but running `type(self)` passes a table instead of the `GuiScript`.

For reference, this is the generated code:
```
function init(self)
    _hxdefold_tmp = _G._hxdefold_self_
    _G._hxdefold_self_ = self
    _hxdefold_.gui_Main_init()
    _G._hxdefold_self_ = _hxdefold_tmp
end
```

Here `self` is saved into `_G._hxdefold_self_`, but the actual `self` instance is lost because `gui_Main_init` calls `script:init`. What's the solution to this problem?

EDIT:
I tried using this little wrapper method to hide the `untyped __lua__` and am disgusted, but it works. I feel like I have to say that I know the naming is terrible because it's almost embarrassing to post it.
```
class Druidd {
	public static function neww():DruidInstance {
		return Druid.neww(untyped __lua__('_G._hxdefold_self_'));
	}
}
```

**Comment by haath** [comment]
@DavidLyhedDanielsson right, this is indeed an issue with the new way in which I've made the `self` reference available to the whole script.

I'd say your approach to passing `_G._hxdefold_self_` is the right one, but it's not ideal since until this moment I considered this global variable to be a hidden implementation detail that users shouldn't ever have to reach for.
But if libraries like Druid need to be given a reference to `self`, then hxdefold should make this field available.
So thanks for this, next time I work on the library I'll most likely add something like a read-only property to scripts called `self` which points to `_G._hxdefold_self_`. I'll make it have a unique abstract type like `SelfReference` and then your druid constructor can expect that as an argument.

Now regarding your callback, it is also true that the current way of handling `self` also complicates callbacks.
What we do now is we capture the callbacks passed to Defold, and switch out the `_G._hxdefold_self_` before and after the callback provided from Haxe. This way any Haxe code ran as a result of the callback will have the correct self pointer to access its fields.
See for example:

https://github.com/hxdefold/hxdefold/blob/8ee24ec921b27fead566534c06fb1b8fd35a4777/src/defold/Http.hx#L31-L42

So when you're writing an extern that involves callbacks, you need to do something similar.

I agree that it's a bit ugly, but it's something that you need to write only once and then it can stay hidden.
As opposed to the old way of carrying `self` with you everywhere, which was boilerplate that you had to repeat over and over.
